sound/soc/sof/sof-audio.c

```diff
@@ -380,49 +396,34 @@ int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, in
 			dev_err(sdev->dev, "error: no pipeline widget found for %s\n",
 				swidget->widget->name);
 			ret = -EINVAL;
-			goto pipe_free;
+			goto widget_free;
 		}
 
-		ret = sof_widget_setup(sdev, pipe_widget);
-		if (ret < 0)
-			goto pipe_free;
-
-		num_pipe_widgets++;
-
 		if (pipe_widget->complete)
 			continue;
 
 		pipe_widget->complete = snd_sof_complete_pipeline(sdev->dev, pipe_widget);
 		if (pipe_widget->complete < 0) {
 			ret = pipe_widget->complete;
-			goto pipe_free;
+			goto widget_free;
 		}
 	}
 
 	return 0;
 
-pipe_free:
-	/* free the pipe widgets that were set up successfully */
+widget_free:
```

Comment:
@ranj063 a little bit concern about the error path here, we have 5 failure cases and will jump to this same label, is the unified handling below can handle all these 5 failures correctly? Does the num_widgets have enough information for us? If no, we might hit swidget->use_count imbalanced issue?

Reply:
@keyonjie yes, the same error path is good enough for all failures. num_widgets represents the number of widgets that are set up successfully and in case of ay error, only those set up successfully need to be freed.